src/wallet/wallet2.cpp

```diff
@@ -1459,24 +1460,30 @@ void wallet2::update_pool_state()
   MDEBUG("update_pool_state done first loop");
 
   // remove pool txes to us that aren't in the pool anymore
-  std::unordered_map<crypto::hash, wallet2::payment_details>::iterator uit = m_unconfirmed_payments.begin();
-  while (uit != m_unconfirmed_payments.end())
+  // but only if we just refreshed, so that the tx can go in
+  // the in transfers list instead (or nowhere if it just
+  // disappeared without being mined)
+  if (refreshed)
   {
-    const crypto::hash &txid = uit->second.m_tx_hash;
-    bool found = false;
-    for (const auto &it2: res.tx_hashes)
+    std::unordered_map<crypto::hash, wallet2::payment_details>::iterator uit = m_unconfirmed_payments.begin();
+    while (uit != m_unconfirmed_payments.end())
```

Comment:
Sorry created a bit confusion yet again, `m_unconfirmed_payments` appears to only use payment id's as keys. The confusion [stemmed from this line](https://github.com/monero-project/monero/blob/master/src/wallet/wallet2.cpp#L1493) which is doing a lookup by tx id on that object. `m_unconfirmed_payments` is probably better off having tx id as the key and payment id in the value portion - I cannot find a lookup by payment id anywhere.

Also, I _think_ [this line](https://github.com/monero-project/monero/blob/master/src/wallet/wallet2.cpp#L1497) should be a find since it is comparing against `.first` which is the map key.

`process_unconfirmed` should be fine (appears so at a glance).